### ISSUE TITLE
pvr.iptvsimple streams do not call xbmc.Player.onAVChange()

### ISSUE THREAD

**Issue by KasparWinckler** [kodi-pvr/pvr.iptvsimple#866]
Hello,

I have noticed that pvr.iptvsimple streams do not call xbmc.Player.onAVChange().

This can be demonstrated with the code below:
```
import xbmc
import xbmcaddon

ADDON = xbmcaddon.Addon()
ADDON_ID = ADDON.getAddonInfo("id")


def log(message, loginfo=True):
    xbmc.log(f"{ADDON_ID} - {message}", xbmc.LOGINFO if loginfo else xbmc.LOGDEBUG)


class Player(xbmc.Player):
    def __init__(self):
        super().__init__()
        self.file = ""

    def get_file(self):
        if self.isPlaying():
            file = self.getPlayingFile()
        else:
            file = ""
        if file != self.file:
            log(f"File change from '{self.file}' to '{file}'")
            self.file = file

    def onAVChange(self):
        file = self.getPlayingFile()
        log(f"onAVChange to '{file}'")


if __name__ == "__main__":
    log("Service started")
    monitor = xbmc.Monitor()
    player = Player()
    while not monitor.waitForAbort(1):
        player.get_file()
    log("Service stopped")

```

Here is a log when I start a pvr.hts stream (pvr file), then a pvr.iptvsimple stream (aac file), then the pvr.hts stream, then the pvr.iptvsimple stream:
```
2024-05-26 11:54:53.202 T:409924    info <general>: service.onavchange - Service started
2024-05-26 11:55:06.903 T:409924    info <general>: service.onavchange - onAVChange to 'pvr://channels/radio/All%20channels@-1/1@pvr.hts_219690937.pvr'
2024-05-26 11:55:07.003 T:409924    info <general>: service.onavchange - onAVChange to 'pvr://channels/radio/All%20channels@-1/1@pvr.hts_219690937.pvr'
2024-05-26 11:55:07.202 T:409924    info <general>: service.onavchange - File change from '' to 'pvr://channels/radio/All%20channels@-1/1@pvr.hts_219690937.pvr'
2024-05-26 11:55:07.302 T:409924    info <general>: service.onavchange - onAVChange to 'pvr://channels/radio/All%20channels@-1/1@pvr.hts_219690937.pvr'
2024-05-26 11:55:25.203 T:409924    info <general>: service.onavchange - File change from 'pvr://channels/radio/All%20channels@-1/1@pvr.hts_219690937.pvr' to 'https://icecast.radiofrance.fr/fip-hifi.aac'
2024-05-26 11:55:42.204 T:409924    info <general>: service.onavchange - File change from 'https://icecast.radiofrance.fr/fip-hifi.aac' to ''
2024-05-26 11:55:42.605 T:409924    info <general>: service.onavchange - onAVChange to 'pvr://channels/radio/All%20channels@-1/1@pvr.hts_219690937.pvr'
2024-05-26 11:55:42.706 T:409924    info <general>: service.onavchange - onAVChange to 'pvr://channels/radio/All%20channels@-1/1@pvr.hts_219690937.pvr'
2024-05-26 11:55:43.204 T:409924    info <general>: service.onavchange - File change from '' to 'pvr://channels/radio/All%20channels@-1/1@pvr.hts_219690937.pvr'
2024-05-26 11:56:30.207 T:409924    info <general>: service.onavchange - File change from 'pvr://channels/radio/All%20channels@-1/1@pvr.hts_219690937.pvr' to 'https://icecast.radiofrance.fr/fip-hifi.aac'

``` 

This log shows that the pvr.iptvsimple stream does not call xbmc.Player.onAVChange().

I believe that this behaviour is not expected.
I have only observed it with pvr.iptvsimple streams.
I therefore opened this issue here, rather than at Kodi.

Thank you for your help with this.

**Comment by phunkyfish** [comment]
This is entirely down to the inputstream used, either kodi’s, adaptive or ffmpegdirect.

**Comment by phunkyfish** [comment]
So maybe try changing the inputstream and see if that works

**Comment by KasparWinckler** [comment]
Thank you for your answer.

According to the log, the pvr.iptvsimple stream (aac file) is read with inputstream.ffmpegdirect:
```
2024-05-26 20:12:33.066 T:669200    info <general>: AddOnLog: pvr.iptvsimple: pvr.iptvsimple - GetChannelStreamProperties - Live Stream URL: https://icecast.radiofrance.fr/fip-hifi.aac
...
2024-05-26 20:12:34.001 T:409924    info <general>: service.onavchange - File change from '' to 'https://icecast.radiofrance.fr/fip-hifi.aac'
...
2024-05-26 20:12:34.669 T:669133    info <general>: CDVDAudioCodecFFmpeg::Open() Successful opened audio decoder aac
```

I do not know how to change the inputstream for an .aac file in the settings of pvr.iptvsimple or in the M3U playlist. Is it possible?

Should I open an issue at inputstream.ffmpegdirect?

**Comment by phunkyfish** [comment]
Why is this required?

**Comment by KasparWinckler** [comment]
I want to switch audio output from TV for video, to HiFi for audio.
I wrote an addon to do this, that expected xbmc.Player.onAVChange() to detect stream changes.
It worked fine with pvr.hts, but I had to migrate from pvr.hts to pvr.iptvsimple because of issues with ffmpeg 6.0.
I found out that xbmc.Player.onAVChange() does not detect streams started with pvr.iptvsimple+inputstream.ffmpegdirect.
I updated my addon to rely on a loop like in the code in the first comment, rather than xbmc.Player.onAVChange().
I nevertheless believe that it would be better that xbmc.Player.onAVChange() behaves as documented and expected.
Thank you for your help with this.
